### PR TITLE
Implemented separate update configuration callback for bindings with …

### PR DESCRIPTION
**Pull request by dnobel**
…new REST endpoint.

Bug: 471050
Signed-off-by: Dennis Nobel <d.nobel@external.telekom.de>

### CODE REVIEW

**Comment by dnobel** [comment]
@cdjackson This PR "fixes" https://bugs.eclipse.org/bugs/show_bug.cgi?id=471050. Hope that fits to what we discussed here https://github.com/eclipse/smarthome/pull/327 and it fulfills your requirements for the z-wave binding.

**Comment by dnobel** [comment]
In this PR I also fixed a minor issue with the remove handling, pulled the new methods (remove and forceRemove) up to the interface and added options to the REST interface to do a force remove.

File: bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java [hunk 1]
```diff
@@ -173,19 +186,6 @@ public void thingUpdated(Thing thing) {
         initialize();
     }
 
-    public void configurationChanged(Map<String, String> changedParameters) throws Exception {
```
Comment (dnobel) [inline bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/binding/BaseThingHandler.java]: This method came accidentally in by another PR. It should never had made it into ESH, sorry for that. But I guess no one used it.

**Comment by dnobel** [comment]
Configuration updates can now be sent via `PUT /rest/things/{thingUID}/config`.

In the body one has to provide a simple JSON Object:

```javascript
{
   param1: 'value',
   param2: true,
   param3: 20.3
}
```

Old configuration mechanism is still working and everything is backward-compatible. Both at UI/REST side, as well as in the bindings.

**Comment by kaikreuzer** [comment]
Looks all good to me.
Waiting for @cdjackson's feedback before I merge the PR.

**Comment by cdjackson** [comment]
> On 17 Jul 2015, at 09:43, Dennis Nobel <notifications@github.com> wrote:
> 
> @cdjackson <https://github.com/cdjackson> This PR "fixes" https://bugs.eclipse.org/bugs/show_bug.cgi?id=471050 <https://bugs.eclipse.org/bugs/show_bug.cgi?id=471050>. Hope that fits to what we discussed here #327 <https://github.com/eclipse/smarthome/pull/327> and it fulfills your requirements for the z-wave binding.

Thanks Dennis. I’ve not looked at it in detail yet, but from the description on bugzilla it sounds like it should do the trick.

Cheers
Chris

**Comment by cdjackson** [comment]
What do I need to do to get this into OH2? I’ve tried updating the target, but it’s still not picked up the new handler…

Also, there’s still a missing part of this equation I think - and that’s the definition. This PR allows actions to be implemented in that we can now send specific commands/config updates to bindings. What we don’t have is a way to define these as actions.

We could use the context, or the group name, but there would have to be some sort of standard to ensure that UIs implemented this in an appropriate way...

**Comment by dnobel** [comment]
Chris, I think we agreed that we don´t need the concept of "real actions" for the moment:

> For real device actions/functions (device functionality, which does not have a value like a "reset" function), we would like you to ask if we could shift this feature to the future. For your current z-wave binding you could model it with a configuration parameter or a channel. Moreover you could also use a specific "hidden" parameter, which only HABmins knows, for example (if the changedParameters map contains a value "reset", you can trigger the reset). 

According to our discussion, one of the biggest problem was, that you didn´t had a callback where the submitted values of the UI can be received. This should be solved now.

If you want to model an action like a reset for the moment, I would suggest to do it with a boolean property and render a switch or using a channel.

**Comment by cdjackson** [comment]
Hi Dennis,
Absolutely - the biggest issue is solved (thanks), and we also agreed that we didn't need 'real actions'. I can easily use what we have now, but was wondering if we shouldn't try and standardise on something? Currently I'm doing two things - setting the context to 'menu' and putting all these into a certain category (called 'tools' I think). In HABmin I'll then treat these differently, so from that perspective it's fine...

However PaperUI will treat these differently, as will other UIs - do we need to standardise? If not, then we will end up with each binding doing its own thing and each UI presenting the information differently.

Cheers
Chris

**Pull request merged**